src/fft.rs

```diff
@@ -0,0 +1,299 @@
+    ($func_name:ident, $precision:ty, $fft_func:ident) => {
+        /// Performs a Real-Valued Fast Fourier Transform (FFT)
+        ///
+        /// This function computes the FFT of a real-valued input signal and produces
+        /// complex-valued output. The implementation follows the principles of splitting
+        /// the input into even and odd components and then performing the FFT on these
+        /// components.
+        ///
+        /// # Arguments
+        ///
+        /// `input_re` - A slice containing the real-valued input signal.
+        ///
+        /// `output_re` - A mutable slice to store the real parts of the FFT output.
+        ///
+        /// `output_im` - A mutable slice to store the imaginary parts of the FFT output.
+        ///
+        /// # Panics
+        ///
+        /// Panics if `output_re.len() != output_im.len()` and `input_re.len()` == `output_re.len()`
+        ///
+        /// # Examples
+        ///
+        /// ```
+        /// use phastft::fft::{r2c_fft_f32, r2c_fft_f64};
+        ///
+        /// let big_n = 16;
+        /// let input: Vec<f64> = (1..=big_n).map(|x| x as f64).collect();
+        /// let mut output_re = vec![0.0; big_n];
+        /// let mut output_im = vec![0.0; big_n];
+        /// r2c_fft_f64(&input, &mut output_re, &mut output_im);
+        ///
+        /// let input: Vec<f32> = (1..=big_n).map(|x| x as f32).collect();
+        /// let mut output_re: Vec<f32> = vec![0.0; 16];
+        /// let mut output_im: Vec<f32> = vec![0.0; 16];
+        /// r2c_fft_f32(&input, &mut output_re, &mut output_im);
+        /// ```
+        /// # References
+        ///
+        /// This implementation is based on the concepts discussed in
+        /// [Levente Kovács' post](https://kovleventer.com/blog/fft_real/).
+        pub fn $func_name(
+            input_re: &[$precision],
+            output_re: &mut [$precision],
+            output_im: &mut [$precision],
+        ) {
+            assert!(output_re.len() == output_im.len() && input_re.len() == output_re.len());
+            let big_n = input_re.len();
+
+            // Splitting odd and even
+            let (mut z_even, mut z_odd): (Vec<_>, Vec<_>) =
+                input_re.chunks_exact(2).map(|c| (c[0], c[1])).unzip();
+
+            // Z = np.fft.fft(z)
+            $fft_func(&mut z_even, &mut z_odd, Direction::Forward);
+
+            let mut z_x_re = Vec::with_capacity(z_even.len());
+            let mut z_x_im = Vec::with_capacity(z_even.len());
+            let mut z_y_re = Vec::with_capacity(z_even.len());
+            let mut z_y_im = Vec::with_capacity(z_even.len());
+
+            z_x_re.push(0.0);
+            z_x_im.push(0.0);
+            z_y_re.push(0.0);
+            z_y_im.push(0.0);
+
+            // Zminconj = np.roll(np.flip(Z), 1).conj()
+            // Zx =  0.5  * (Z + Zminconj)
+            // Zy = -0.5j * (Z - Zminconj)
+            z_even
+                .iter()
+                .skip(1)
```

Comment:
`.skip(1)` may sometimes give worse codegen than plain `iter()` over `slice[1..]` so that's something to potentially look into.